In [7]:
from sklearn.datasets import fetch_california_housing
import pandas as pd

# Load the California housing dataset
housing = fetch_california_housing(as_frame=True)
df_california = housing.frame

# Display the first few rows of the new dataset
print("California Housing Dataset Head:")
display(df_california.head())

California Housing Dataset Head:


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


In [9]:
# Display info about the new dataset
print("\nCalifornia Housing Dataset Info:")
df_california.info()


California Housing Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


In [10]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error

# Define features (X) and target (y) for the California Housing Dataset
X_california = df_california.drop('MedHouseVal', axis=1)
y_california = df_california['MedHouseVal']

# Split the data into training and testing sets
X_train_california, X_test_california, y_train_california, y_test_california = train_test_split(X_california, y_california, test_size=0.2, random_state=42)

# Scale the features
scaler_california = StandardScaler()
X_train_scaled_california = scaler_california.fit_transform(X_train_california)
X_test_scaled_california = scaler_california.transform(X_test_california)

# Train and evaluate Linear Regression
lr_california = LinearRegression()
lr_california.fit(X_train_scaled_california, y_train_california)
lr_pred_california = lr_california.predict(X_test_scaled_california)
mselr_california = mean_squared_error(y_test_california, lr_pred_california)
lrr2_california = r2_score(y_test_california, lr_pred_california)

print(f"California Housing - Linear Regression MSE: {mselr_california}")
print(f"California Housing - Linear Regression R2: {lrr2_california}")

California Housing - Linear Regression MSE: 0.5558915986952442
California Housing - Linear Regression R2: 0.575787706032451


In [12]:
from sklearn.ensemble import RandomForestRegressor

# Define features (X) and target (y) for the California Housing Dataset
X_california = df_california.drop('MedHouseVal', axis=1)
y_california = df_california['MedHouseVal']

# Split the data into training and testing sets (using the same split as Linear Regression)
X_train_california, X_test_california, y_train_california, y_test_california = train_test_split(X_california, y_california, test_size=0.2, random_state=42)

# Initialize the Random Forest Regressor model with 100 estimators
RFmodel = RandomForestRegressor(n_estimators=100, random_state=42)

# Train the Random Forest model using the scaled training data
# Note: The original X_train_scaled_california from the previous cell is used here.
RFmodel.fit(X_train_scaled_california, y_train_california)

# Make predictions on the scaled test data
# Note: The original X_test_scaled_california from the previous cell is used here.
RFpred = RFmodel.predict(X_test_scaled_california)

# Calculate Mean Squared Error (MSE)
mseRF = mean_squared_error(y_test_california, RFpred)

# Calculate R-squared (R2) score
r2RF = r2_score(y_test_california, RFpred)

# Print the evaluation metrics
print(f"California Housing - Random Forest MSE: {mseRF}")
print(f"California Housing - Random Forest R2: {r2RF}")

California Housing - Random Forest MSE: 0.255169737347244
California Housing - Random Forest R2: 0.8052747336256919


## Model Performance Insights

Based on the training and evaluation of the Linear Regression and Random Forest models on the California Housing dataset, here are the key insights:

*   **Linear Regression Model:**
    *   Mean Squared Error (MSE): `0.556`
    *   R-squared (R2) Score: `0.576`

*   **Random Forest Regressor Model:**
    *   Mean Squared Error (MSE): `0.255`
    *   R-squared (R2) Score: `0.805`

**Conclusion:**

The Random Forest Regressor model significantly outperforms the Linear Regression model on this dataset. It achieves a much lower Mean Squared Error (indicating closer predictions to actual values) and a substantially higher R-squared score (meaning it explains a greater proportion of the variance in the target variable).

This suggests that the relationship between the features and the median house value in the California housing dataset is likely non-linear and complex, which the Random Forest model is better equipped to capture compared to a simpler linear model.